pkg/cmd/cli/cmd/rsync.go

```diff
@@ -0,0 +1,372 @@
+	}
+	if !o.UseTar && len(o.RshCommand) == 0 {
+		return fmt.Errorf("rsh command must be provided when not using tar")
+	}
+	if o.Out == nil || o.ErrOut == nil {
+		return fmt.Errorf("output and error streams must be specified")
+	}
+	if o.LocalExecutor == nil || o.RemoteExecutor == nil {
+		return fmt.Errorf("local and remote executors must be provided")
+	}
+	return nil
+}
+
+func (o *RsyncOptions) copyWithRsync() error {
+	glog.V(3).Infof("Copying files with rsync")
+	flags := "-a"
+	if !o.Quiet {
+		flags += "v"
+	}
+	destination := fmt.Sprintf("%s:%s", o.PodName, o.Destination)
+	cmd := []string{defaultRsyncExecutable, flags}
+	if o.Delete {
+		cmd = append(cmd, "--delete")
+	}
+	cmd = append(cmd, "-e", o.RshCommand, o.Source, destination)
+	glog.V(4).Infof("Local command: %s", strings.Join(cmd, " "))
+	return o.LocalExecutor.Execute(cmd, nil, o.Out, o.ErrOut)
+}
+
+func (o *RsyncOptions) copyWithTar() error {
+	glog.V(3).Infof("Copying files with tar")
+	if o.Delete {
+		deleteCmd := []string{"sh", "-c", fmt.Sprintf("rm -rf %s", filepath.Join(o.Destination, "*"))}
+		err := executeWithLogging(o.RemoteExecutor, deleteCmd)
+		if err != nil {
+			return fmt.Errorf("unable to delete files in destination: %v", err)
+		}
+	}
+	tmp, err := ioutil.TempFile("", "rsync")
+	if err != nil {
+		return err
+	}
+	err = tarLocal(o.Tar, o.Source, tmp)
+	if err != nil {
+		return err
+	}
+	err = tmp.Close()
+	if err != nil {
+		return err
+	}
+	tmp, err = os.Open(tmp.Name())
+	if err != nil {
+		return err
+	}
+	defer os.Remove(tmp.Name())
+
+	flags := "x"
+	if !o.Quiet {
+		flags += "v"
+	}
+	remoteCmd := []string{defaultTarExecutable, flags, "-C", o.Destination}
+	return o.RemoteExecutor.Execute(remoteCmd, tmp, o.Out, o.ErrOut)
+}
+
+func tarLocal(tar tar.Tar, sourceDir string, w io.Writer) error {
+	glog.V(4).Infof("Tarring %s locally", sourceDir)
+	// includeParent mimics rsync's behavior. When the source path ends in a path
+	// separator, then only the contents of the directory are copied. Otherwise,
+	// the directory itself is copied.
+	includeParent := true
+	if strings.HasSuffix(sourceDir, string(filepath.Separator)) {
+		includeParent = false
+		sourceDir = sourceDir[:len(sourceDir)-1]
+	}
+	return tar.CreateTarStream(sourceDir, includeParent, w)
+}
+
+// RunRsync copies files from source to destination
+func (o *RsyncOptions) RunRsync() error {
+
+	if !o.UseTar && o.checkLocalRsync() == nil {
+		// If rsync is present in the pod's container, then use it
+		if o.checkRemoteRsync() == nil {
+			return o.copyWithRsync()
+		}
+	}
+
+	// If tar is present in the pod's container, then use it
+	if o.checkRemoteTar() == nil {
+		if !o.UseTar {
+			glog.Warningf("Could not find rsync. It must be on the path of the local machine and of the remote container, using tar.")
+		}
+		return o.copyWithTar()
+	}
+
+	return fmt.Errorf("the container in pod %s/%s does not have rsync or tar", o.Namespace, o.PodName)
+}
```

Comment:
Yeah, even though we will have to parse the arguments in the rsync pattern,
it's more correct and opens up future scenarios.  And it's OK to fetch the
pod info.

On Oct 2, 2015, at 1:07 AM, Cesar Wong <notifications@github.com> wrote:

In pkg/cmd/cli/cmd/rsync.go
<https://github.com/openshift/origin/pull/4787#discussion_r40977859>:

> +		// If rsync is present in the pod's container, then use it
> +		if o.checkRemoteRsync() == nil {
> +			return o.copyWithRsync()
> +		}
> +	}
> +
> +	// If tar is present in the pod's container, then use it
> +	if o.checkRemoteTar() == nil {
> +		if !o.UseTar {
> +			glog.Warningf("Could not find rsync. It must be on the path of the local machine and of the remote container, using tar.")
> +		}
> +		return o.copyWithTar()
> +	}
> +
> +	return fmt.Errorf("the container in pod %s/%s does not have rsync or tar", o.Namespace, o.PodName)
> +}

We need an additional call to figure out what the container name is, is
that ok?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4787/files#r40977859>.